File: Data Structures & Algorithms/house-robber/submission-3.java
class Solution {
    public int[] memo ;
    public int dfs(int i , int[] nums){
        if(i>=nums.length)
            return 0 ;
        if(memo[i]!=0)
            return memo[i];
        memo[i]=Math.max(dfs(i+1,nums),nums[i]+dfs(i+2,nums));
        return memo[i];
    }
    public int rob(int[] nums) {
        memo = new int[nums.length];
        return dfs(0,nums);
    }
}

File: Data Structures & Algorithms/stone-game/submission-3.py
class Solution:
    def stoneGame(self, piles: List[int]) -> bool:
        memo={}
        def dfs(l, r):
            if (l,r) in memo :
                return memo[(l,r)]
            if l > r:
                return 0
            even = (r - l) % 2 == 0
            left = piles[l] if even else 0
            right = piles[r] if even else 0
            memo[(l,r)]= max(dfs(l + 1, r) + left, dfs(l, r - 1) + right)
            return memo[(l,r)]

        total = sum(piles)
        alice_score = dfs(0, len(piles) - 1)
        return alice_score > total - alice_score
